### PR TITLE
Disable ActiveJob retry jitter when given zero/falsey value

### PR DESCRIPTION
**Pull request by cpruitt**
This PR updates the `ActiveJob` `retry_on` and `determine_delay` methods to accept `nil` / `falsey` values for `jitter:` and treat them as `0.0` which effectively disables jitter randomization.

This also fixes the https://github.com/rails/rails/issues/37990 bug which caused a random fraction of a second to be added to the delay when a jitter value of `0.0` was used.

### Refactoring `determine_delay`

This PR refactors `determine_delay` a bit. It combines the `ActiveSupport::Duration` and `Integer` cases as they were identical with the exception of variable names and a call to `.to_i` which is a no-op for `Integer`. If this is undesirable for any reason I can break it back apart.

This also pulls calculation for jitter out into a private `determine_jitter_for_delay(delay, jitter)` method.

Thanks @kaspth for posting code for a starting point in https://github.com/rails/rails/pull/37923#discussion_r357445670

cc @eileencodes for visibility

### CODE REVIEW

File: activejob/lib/active_job/exceptions.rb [hunk 1]
```diff
@@ -126,17 +126,20 @@ def retry_job(options = {})
...
-        when Integer
-          seconds = seconds_or_duration_or_algorithm
-          seconds + (Kernel.rand(seconds * jitter).ceil)
+          delay = executions**4
+          delay_jitter = determine_jitter_for_delay(delay, jitter)
+          delay + delay_jitter + 2
+        when ActiveSupport::Duration, Integer
+          delay = seconds_or_duration_or_algorithm.to_i
+          delay_jitter = determine_jitter_for_delay(delay, jitter)
+          delay + delay_jitter
```
Comment (cpruitt) [inline activejob/lib/active_job/exceptions.rb]: Two `when` cases were combined here. They were extremely similar before so I'm not sure if there's a style preference in Rails for keeping cases to a single condition. Happy to change this back if so.

File: activejob/lib/active_job/exceptions.rb [hunk 2]
```diff
@@ -126,17 +126,20 @@ def retry_job(options = {})
     end
 
     private
-      def determine_delay(seconds_or_duration_or_algorithm:, executions:, jitter: nil)
-        jitter ||= self.class.retry_jitter
+      JITTER_DEFAULT = Object.new
```
Comment (rafaelfranca) [inline activejob/lib/active_job/exceptions.rb]: This is not really private if you don't call `private_constant`, so it is better to move it to the public section or keep it here and also call `private_constant`.

File: activejob/lib/active_job/exceptions.rb [hunk 3]
```diff
@@ -145,6 +148,11 @@ def determine_delay(seconds_or_duration_or_algorithm:, executions:, jitter: nil)
         end
       end
 
+      def determine_jitter_for_delay(delay, jitter)
+        return 0.0 if jitter == 0
```
Comment (rafaelfranca) [inline activejob/lib/active_job/exceptions.rb]:
```suggestion
        return 0.0 if jitter.zero?
```

File: activejob/lib/active_job/exceptions.rb (hunk 2 above)
Comment (cpruitt) [reply activejob/lib/active_job/exceptions.rb]: Good catch. Thanks. I've made it private in https://github.com/rails/rails/pull/38003/commits/98439f3d5c510cd413fac4e71b8e460343e30cff

File: activejob/lib/active_job/exceptions.rb (hunk 3 above)
Comment (cpruitt) [reply activejob/lib/active_job/exceptions.rb]: 👍 Fixed in https://github.com/rails/rails/pull/38003/commits/fff3cd7b890fcc5f8502062201212f3cee911fb2. Thanks.

**Comment by rafaelfranca** [comment]
The indentation is still incorrect.

**Comment by cpruitt** [comment]
> The indentation is still incorrect.

@rafaelfranca Thanks, I saw & fixed. It only took 40 tries. 🤦🏼‍♂️

File: activejob/test/jobs/retry_job.rb [hunk 4]
```diff
@@ -18,6 +19,7 @@ class CustomDiscardableError < StandardError; end
 
 class RetryJob < ActiveJob::Base
   retry_on DefaultsError
+  retry_on DisabledJitterError, jitter: nil
```
Comment (rafaelfranca) [inline activejob/test/jobs/retry_job.rb]: Should we have a test for `0.0` As well?

Comment (cpruitt) [reply activejob/test/jobs/retry_job.rb]: It's probably prudent. I'll add.

Comment (cpruitt) [reply activejob/test/jobs/retry_job.rb]: Added `test "retry jitter disabled with zero"` https://github.com/rails/rails/pull/38003/files#diff-a7e74ec284b847138992693bd53c4746R173

**Pull request merged**